src/rules/noUnnecessaryTypeAssertionRule.ts

```diff
@@ -92,9 +94,64 @@ class Walker extends Lint.AbstractWalker<string[]> {
 
         const uncastType = this.checker.getTypeAtLocation(node.expression);
         if (uncastType === castType) {
-            this.addFailureAtNode(node, Rule.FAILURE_STRING, node.kind === ts.SyntaxKind.TypeAssertionExpression
+            // In some cases, this is still not enough; the type in the
+            // assertion can actually impact the type in the expression being
+            // asserted. To avoid these false positives, we create a new
+            // ts.Program with the edited file, to see if the assertion is still
+            // unnecessary.
+            // This can be a bit slow, so we still want to guard it by the more
+            // basic check above.
+            const replacement = node.kind === ts.SyntaxKind.TypeAssertionExpression
                 ? Lint.Replacement.deleteFromTo(node.getStart(), node.expression.getStart())
-                : Lint.Replacement.deleteFromTo(node.expression.getEnd(), node.getEnd()));
+                : Lint.Replacement.deleteFromTo(node.expression.getEnd(), node.getEnd());
+            const sourceFile = this.sourceFile;
+            const modifiedHost = ts.createCompilerHost(this.program.getCompilerOptions());
```

Comment:
this won't work if the `Program` was created using a custom `CompilerHost`.
I know we also do this in `Linter.updateProgram`, but that's only executed if you run with `--fix`.